### DISCUSSION TITLE
bug: addForce() is framerate dependent

### DISCUSSION THREAD

**Discussion by themerk** [kaplayjs/kaplay#557] [Q&A] [answered]
## **Describe the bug**

The docs indicate that `addForce()` is in units of `kg*px/s^2`, meaning it shouldn't have to be scaled by `dt()` and should give the same result regardless of the frame rate. Unfortunately, that's not the case. Slowing the framerate reduces the amount of force applied and changes the motion of the object.

Is this a bug or am I misunderstanding how the physics system applies forces?

## **Version**

Replicated in both 3001.0.2 and 4000.0.0-alpha15

## **To Reproduce**

[Example on KAPLAYGROUND](https://play.kaplayjs.com/?code=eJxlUMtuwjAQvPsr9uhICAwSPVDREyD10If6EIeqhyXekAjHazkGEqH%2Be21IEVJvM6OZfcxoJEYjWLZYO0PABaDWK%2FY5yQw2VNktaHJkNdkAbKHwWBN4DBRjKfnhO9BVUZBPjgOaPTVQsIca29XrO0jaDmE8hajcqQzQarAcCEJJNzmHoUzLk7pBY4TYoTPYyZMAOFY6lDOYKjWIrKRqW4YrvayZxeHiJ7sXImfbBGhhnh6RX9EBkFc%2BNyTHKhtcOBv28m256LnjRvYQPeEf3rDuEv5OcwXbT6fj3zIWM3%2BA09kSC3jiQywk3XRW2mEdBTmeqAGoFOxt505B89Ee0eum917LXrysn4dNjvHMiVLZTXDNfoee91b3Skw5Z7rH2u1N8y%2FZQx1klqbETn4Bz7eJKA%3D%3D)

```js
kaplay({
  width: 500,
  height: 500,
  maxFPS: 60  // Try different values for maxFPS (eg. 15 or 60) and note the different path of the ball
});

const x = add([
    circle(10),
    color(RED),
    pos(),
    area(),
    body(),
]);

onUpdate(() => {
    // Move right
    x.move(120, 0);

    // Force downwards
    x.addForce(DOWN.scale(200));
});
```

## **Workaround**

In my app I just use `applyImpulse()` instead, scaling by dt():

```js
x.applyImpulse(DOWN.scale(200).scale(dt()));
```

**Comment by mflerackers** [comment]
What exactly are you trying to do? First of all, physics runs at a fixed 50 fps, regardless of the visual fps.
A force needs to be applied for a while to have an effect, while an impulse is an instantaneous change in velocity. None of these need to be multiplied by dt. Especially since dt is dependent on the visual fps, while the physics dt is fixed, 20ms, so I have no idea why you are trying to mix these.
Lastly, move is not physics based, it just uses dt internally. So you're mixing 2 kinds of simulations. One with fixed integration steps, and one with variable integration steps. You should use either one or the other if you want to get any kind of consistent result.

**Reply by themerk** [reply]
Fair enough, I can see how that's a confusing example. Here's a simpler version that just applies a constant force for 2s and then pauses. The ball should accelerate downwards and stop.

```js
kaplay({
  width: 500,
  height: 500,
  maxFPS: 60
});

const x = add([
    circle(10),
    pos(250, 10),
    body(),
]);

onUpdate(() => {
    // Force downwards
    x.addForce(DOWN.scale(200));
});

// After 3 seconds, pause
wait(2, () => x.paused = true);
```

If you change maxFPS to 30, the ball stops in a different place. Obviously the visual FPS shouldn't be affecting the actual physics result.

I can see now why this is happening. Physics is running at a fixed 50Hz but the visual framerate is changing, and hence `onUpdate()` is being called at different rates. I don't want to apply the force every _visual_ frame, but rather every _physics_ frame.

Is there some equivalent to `onUpdate()` that gets called once per physics frame?

**Reply by dragoncoder047** [reply, marked as answer]
It should be `onFixedUpdate()`... not sure why it isn't coming up in the documentation, the function exists...

```js
onFixedUpdate(() => {
    // Force downwards
    x.addForce(DOWN.scale(200));
});
